File: Count Covered Buildings/Count Covered Buildings.py
# Solution for Count Covered Buildings in PY

class Solution:
    def countCoveredBuildings(self, n: int, buildings: List[List[int]]) -> int:

        # Maps row (y) → (minX, maxX)
        rowToMinMaxX = {}

        # Maps column (x) → (minY, maxY)
        colToMinMaxY = {}

        # Step 1: Precompute boundaries
        for x, y in buildings:
            if y not in rowToMinMaxX:
                rowToMinMaxX[y] = [float('inf'), float('-inf')]
            if x not in colToMinMaxY:
                colToMinMaxY[x] = [float('inf'), float('-inf')]

            # Update row boundaries
            rowToMinMaxX[y][0] = min(rowToMinMaxX[y][0], x)
            rowToMinMaxX[y][1] = max(rowToMinMaxX[y][1], x)

            # Update column boundaries
            colToMinMaxY[x][0] = min(colToMinMaxY[x][0], y)
            colToMinMaxY[x][1] = max(colToMinMaxY[x][1], y)

        # Step 2: Count covered buildings
        covered = 0
        for x, y in buildings:
            minX, maxX = rowToMinMaxX[y]
            minY, maxY = colToMinMaxY[x]

            hasLeft = minX < x
            hasRight = x < maxX
            hasBelow = minY < y
            hasAbove = y < maxY

            if hasLeft and hasRight and hasBelow and hasAbove:
                covered += 1

        return covered

File: Count Covered Buildings/Count Covered Buildings.java
// Solution for Count Covered Buildings in JAVA

class Solution {
    public int countCoveredBuildings(int n, int[][] buildings) {

        // Maps row (y) → {minX, maxX}
        Map<Integer, int[]> rowToMinMaxX = new HashMap<>();

        // Maps column (x) → {minY, maxY}
        Map<Integer, int[]> colToMinMaxY = new HashMap<>();

        // Step 1: Precompute boundaries
        for (int[] b : buildings) {
            int x = b[0];
            int y = b[1];

            rowToMinMaxX.putIfAbsent(y, new int[]{Integer.MAX_VALUE, Integer.MIN_VALUE});
            colToMinMaxY.putIfAbsent(x, new int[]{Integer.MAX_VALUE, Integer.MIN_VALUE});

            // Update row boundaries
            rowToMinMaxX.get(y)[0] = Math.min(rowToMinMaxX.get(y)[0], x);
            rowToMinMaxX.get(y)[1] = Math.max(rowToMinMaxX.get(y)[1], x);

            // Update column boundaries
            colToMinMaxY.get(x)[0] = Math.min(colToMinMaxY.get(x)[0], y);
            colToMinMaxY.get(x)[1] = Math.max(colToMinMaxY.get(x)[1], y);
        }

        // Step 2: Count covered buildings
        int covered = 0;

        for (int[] b : buildings) {
            int x = b[0];
            int y = b[1];

            int[] rowLimits = rowToMinMaxX.get(y); // {minX, maxX}
            int[] colLimits = colToMinMaxY.get(x); // {minY, maxY}

            boolean hasLeft = rowLimits[0] < x;
            boolean hasRight = x < rowLimits[1];
            boolean hasBelow = colLimits[0] < y;
            boolean hasAbove = y < colLimits[1];

            if (hasLeft && hasRight && hasBelow && hasAbove) {
                covered++;
            }
        }

        return covered;
    }
}
